### PR TITLE
[VLM][Model] Support image input for Chameleon

### LINKED ISSUE #5721: [New Model]: Chameleon support
### The model to consider.

https://huggingface.co/facebook/chameleon
(as of now, the models can be downloaded using the [model form](https://ai.meta.com/resources/models-and-libraries/chameleon-downloads/))

Chameleon is an interesting multimodal model architecture based on Llama 2. It adds image inputs and outputs to Llama 2 by tokenizing images using a VQ-VAE and adding the codebook to Llama's tokenizer vocabulary.
In principle, it supports text and images as input and output in arbitrary combination. However, the released models were finetuned to prevent image generation.

### The closest model vllm already supports.

LlamaForCausalLM

### What's your difficulty of supporting the model you want?

For text->text support, the implementation should actually be fairly easy. The model is based on Llama-2 with the following differences:
* QK norm
* reordering the norm similar to Swin Transformer (normalizing the outputs of the attention and ffn blocks instead of the inputs)

To enable image inputs, image tokenization using the provided VQ-VAE needs to be added.

Further info:
* Original paper: https://arxiv.org/abs/2405.09818
* Reference implementation: https://github.com/facebookresearch/chameleon

### PR DESCRIPTION
**Pull request by ywang96**
This PR is a follow-up to #5770 that adds image + text -> text support for Chameleon. Current implementation only allows one image per request.

FIXES #5721 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
It would be great if you could add a test case for inputting images into Chameleon!

**Comment by xwjiang2010** [comment]
Running into an OOM error under default setting. A100 - 80G, 7B Chameleon model.

Code
```
from PIL import Image


from vllm import LLM

def run():
    llm = LLM(model="/mnt/local_storage/data/cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/bbd94872b11573e40a061783cd1766408614b6e4/") # , max_num_seqs=10)
    # llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-30b/snapshots/1c95500025a2a51183c7bb8f410ecd3d84e4cb2c/")

    prompt = "Write me a poetry for kids based on this image."


    image = Image.open("/home/ray/default/0.JPG")
    outputs = llm.generate({
        "prompt": prompt,
        "multi_modal_data": {
            "image": image
        },
    })

    for o in outputs:
        generated_text = o.outputs[0].text
        print(generated_text)

run()
```

Stacktrace
```
rank0]: Traceback (most recent call last):
[rank0]:   File "/home/ray/default/vllm/test_chameleon.py", line 28, in <module>
[rank0]:     run_llava()
[rank0]:   File "/home/ray/default/vllm/test_chameleon.py", line 10, in run_llava
[rank0]:     llm = LLM(model="/mnt/local_storage/data/cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/bbd94872b11573e40a061783cd1766408614b6e4/")
[rank0]:   File "/home/ray/default/vllm/vllm/entrypoints/llm.py", line 156, in __init__
[rank0]:     self.llm_engine = LLMEngine.from_engine_args(
[rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 440, in from_engine_args
[rank0]:     engine = cls(
[rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 264, in __init__
[rank0]:     self._initialize_kv_caches()
[rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 363, in _initialize_kv_caches
[rank0]:     self.model_executor.determine_num_available_blocks())
[rank0]:   File "/home/ray/default/vllm/vllm/executor/gpu_executor.py", line 94, in determine_num_available_blocks
[rank0]:     return self.driver_worker.determine_num_available_blocks()
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/worker/worker.py", line 179, in determine_num_available_blocks
[rank0]:     self.model_runner.profile_run()
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/worker/model_runner.py", line 767, in profile_run
[rank0]:     self.execute_model(model_input, kv_caches, intermediate_tensors)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/worker/model_runner.py", line 1185, in execute_model
[rank0]:     hidden_or_intermediate_states = model_executable(
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 946, in forward
[rank0]:     image_tokens = self.model.get_image_tokens(image_input["data"].to(
[rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 843, in get_image_tokens
[rank0]:     _, _, image_toks = self.vqmodel.encode(pixel_values)
[rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 741, in encode
[rank0]:     hidden_states = self.encoder(pixel_values)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 709, in forward
[rank0]:     hidden_states.append(self.down[i_level].downsample(
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 487, in forward
[rank0]:     hidden_states = self.conv(hidden_states)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 460, in forward
[rank0]:     return self._conv_forward(input, self.weight, self.bias)
[rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 456, in _conv_forward
[rank0]:     return F.conv2d(input, weight, bias, self.stride,
[rank0]: torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 4.00 GiB. GPU 
```

I think there is something wrong with the profiling logic. Especially considering the hardware is pretty generous for the size of the model run.

**Comment by ywang96** [comment]
> Running into an OOM error under default setting. A100 - 80G, 7B Chameleon model.
> 
> Code
> 
> ```
> from PIL import Image
> 
> 
> from vllm import LLM
> 
> def run():
>     llm = LLM(model="/mnt/local_storage/data/cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/bbd94872b11573e40a061783cd1766408614b6e4/") # , max_num_seqs=10)
>     # llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-30b/snapshots/1c95500025a2a51183c7bb8f410ecd3d84e4cb2c/")
> 
>     prompt = "Write me a poetry for kids based on this image."
> 
> 
>     image = Image.open("/home/ray/default/0.JPG")
>     outputs = llm.generate({
>         "prompt": prompt,
>         "multi_modal_data": {
>             "image": image
>         },
>     })
> 
>     for o in outputs:
>         generated_text = o.outputs[0].text
>         print(generated_text)
> 
> run()
> ```
> 
> Stacktrace
> 
> ```
> rank0]: Traceback (most recent call last):
> [rank0]:   File "/home/ray/default/vllm/test_chameleon.py", line 28, in <module>
> [rank0]:     run_llava()
> [rank0]:   File "/home/ray/default/vllm/test_chameleon.py", line 10, in run_llava
> [rank0]:     llm = LLM(model="/mnt/local_storage/data/cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/bbd94872b11573e40a061783cd1766408614b6e4/")
> [rank0]:   File "/home/ray/default/vllm/vllm/entrypoints/llm.py", line 156, in __init__
> [rank0]:     self.llm_engine = LLMEngine.from_engine_args(
> [rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 440, in from_engine_args
> [rank0]:     engine = cls(
> [rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 264, in __init__
> [rank0]:     self._initialize_kv_caches()
> [rank0]:   File "/home/ray/default/vllm/vllm/engine/llm_engine.py", line 363, in _initialize_kv_caches
> [rank0]:     self.model_executor.determine_num_available_blocks())
> [rank0]:   File "/home/ray/default/vllm/vllm/executor/gpu_executor.py", line 94, in determine_num_available_blocks
> [rank0]:     return self.driver_worker.determine_num_available_blocks()
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
> [rank0]:     return func(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/worker/worker.py", line 179, in determine_num_available_blocks
> [rank0]:     self.model_runner.profile_run()
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
> [rank0]:     return func(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/worker/model_runner.py", line 767, in profile_run
> [rank0]:     self.execute_model(model_input, kv_caches, intermediate_tensors)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
> [rank0]:     return func(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/worker/model_runner.py", line 1185, in execute_model
> [rank0]:     hidden_or_intermediate_states = model_executable(
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
> [rank0]:     return self._call_impl(*args, **kwargs)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
> [rank0]:     return forward_call(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 946, in forward
> [rank0]:     image_tokens = self.model.get_image_tokens(image_input["data"].to(
> [rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 843, in get_image_tokens
> [rank0]:     _, _, image_toks = self.vqmodel.encode(pixel_values)
> [rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 741, in encode
> [rank0]:     hidden_states = self.encoder(pixel_values)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
> [rank0]:     return self._call_impl(*args, **kwargs)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
> [rank0]:     return forward_call(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 709, in forward
> [rank0]:     hidden_states.append(self.down[i_level].downsample(
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
> [rank0]:     return self._call_impl(*args, **kwargs)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
> [rank0]:     return forward_call(*args, **kwargs)
> [rank0]:   File "/home/ray/default/vllm/vllm/model_executor/models/chameleon.py", line 487, in forward
> [rank0]:     hidden_states = self.conv(hidden_states)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
> [rank0]:     return self._call_impl(*args, **kwargs)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
> [rank0]:     return forward_call(*args, **kwargs)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 460, in forward
> [rank0]:     return self._conv_forward(input, self.weight, self.bias)
> [rank0]:   File "/home/ray/anaconda3/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 456, in _conv_forward
> [rank0]:     return F.conv2d(input, weight, bias, self.stride,
> [rank0]: torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 4.00 GiB. GPU 
> ```
> 
> I think there is something wrong with the profiling logic. Especially considering the hardware is pretty generous for the size of the model run.

Thank you for catching this - this has been fixed!

File: vllm/model_executor/models/chameleon.py [hunk 1]
```diff
@@ -318,12 +426,333 @@ def forward(
...
+        last_hidden_state = self.norm_out(last_hidden_state)
+        last_hidden_state *= torch.sigmoid(last_hidden_state)
+        last_hidden_state = self.conv_out(last_hidden_state)
+        return last_hidden_state
+
+
+# Adapted from transformers.models.chameleon.modeling_chameleon.ChameleonVQVAE #noqa
+class ChameleonVQVAE(nn.Module):
+    config_class = ChameleonVQVAEConfig
+    _no_split_modules = ["ChameleonVQVAEVectorQuantizer"]
```
Comment (xwjiang2010) [inline vllm/model_executor/models/chameleon.py:730]: what is this for?

File: vllm/model_executor/models/chameleon.py [hunk 2]
```diff
@@ -318,12 +426,333 @@ def forward(
...
+        super().__init__()
+        self.encoder = ChameleonVQVAEEncoder(config)
+        self.quantize = ChameleonVQVAEVectorQuantizer(config)
+        self.quant_conv = torch.nn.Conv2d(config.latent_channels,
+                                          config.embed_dim, 1)
+        self.post_quant_conv = torch.nn.Conv2d(config.embed_dim,
+                                               config.latent_channels, 1)
+        self.eval()  # Chameleon's VQ model is frozen
+
+    def encode(self, pixel_values: torch.Tensor):
```
Comment (xwjiang2010) [inline vllm/model_executor/models/chameleon.py:742]: add return types.

File: vllm/model_executor/models/chameleon.py [hunk 3]
```diff
@@ -401,13 +830,23 @@ def __init__(
...
         self.norm = RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
-
-        # TODO: Support image input
-        # self.vqmodel = ChameleonVQModel(config.vq_config)
+        self.vqmodel = ChameleonVQVAE(config.vq_config)
 
     def get_input_embeddings(self, input_ids: torch.Tensor) -> torch.Tensor:
         return self.embed_tokens(input_ids)
 
+    def get_image_tokens(self, pixel_values: torch.Tensor):
```
Comment (xwjiang2010) [inline vllm/model_executor/models/chameleon.py:838]: return types

File: vllm/transformers_utils/configs/chameleon.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def __init__(
         rope_scaling=None,
         attention_bias=False,
         attention_dropout=0.0,
-        qk_layernorm=False,
```
Comment (xwjiang2010) [inline vllm/transformers_utils/configs/chameleon.py:34]: this is not a thing anymore?

**Comment by xwjiang2010** [comment]
I am done with manual testing.
Understand that we cannot add correctness test now. But can we have one test that exercises the code? Preferably with multiple images in a batch.

File: vllm/model_executor/models/chameleon.py (hunk 1 above)
Comment (ywang96) [reply vllm/model_executor/models/chameleon.py:730]: This was just copied from transformers - will remove for code cleanness

File: vllm/transformers_utils/configs/chameleon.py (hunk 4 above)
Comment (ywang96) [reply vllm/transformers_utils/configs/chameleon.py:34]: In the latest version they removed this config and always assume q_norm and k_norm needs to be applied

**Comment by ywang96** [comment]
I have addressed all comments and added a batched test! cc @xwjiang2010 @DarkLight1337

**Review by DarkLight1337 — APPROVED** [review]
Thanks for implementing this!

**Pull request merged**